fish-rust/src/function.rs

```diff
@@ -0,0 +1,711 @@
+// Functions for storing and retrieving function information. These functions also take care of
+// autoloading functions in the $fish_function_path. Actual function evaluation is taken care of by
+// the parser and to some degree the builtin handling library.
+
+use crate::ast::{self, Node};
+use crate::common::{escape, valid_func_name, FilenameRef};
+use crate::env::{EnvStack, Environment};
+use crate::event::{self, EventDescription};
+use crate::ffi::{self, parser_t, Repin};
+use crate::global_safety::RelaxedAtomicBool;
+use crate::parse_tree::{NodeRef, ParsedSourceRefFFI};
+use crate::parser_keywords::parser_keywords_is_reserved;
+use crate::wchar::{wstr, WString, L};
+use crate::wchar_ext::WExt;
+use crate::wchar_ffi::wcstring_list_ffi_t;
+use crate::wchar_ffi::{AsWstr, WCharFromFFI, WCharToFFI};
+use crate::wutil::{dir_iter::DirIter, gettext::wgettext_expr, sprintf};
+use cxx::{CxxWString, UniquePtr};
+use once_cell::sync::Lazy;
+use std::collections::{HashMap, HashSet};
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+
+#[derive(Clone)]
+pub struct FunctionProperties {
+    /// Reference to the node, along with the parsed source.
+    pub func_node: NodeRef<ast::BlockStatement>,
+
+    /// List of all named arguments for this function.
+    pub named_arguments: Vec<WString>,
+
+    /// Description of the function.
+    pub description: WString,
+
+    /// Mapping of all variables that were inherited from the function definition scope to their
+    /// values.
+    pub inherit_vars: HashMap<WString, Vec<WString>>,
```

Comment:
TODO: It would be much cheaper memory-wise to use a sorted vec rather than a `HashMap` here (this is per function!), and probably faster too. (How many variables does the typical function which shadows a variable actually end up shadowing?)

An empty vec is alloc-free but I think the same isn't true for an empty `HashMap`.

Reply:
Good suggestion. I am pretty sure an empty HashMap does not allocate but it's still [gross](https://rust.godbolt.org/z/76YanTPPo). I didn't know there was TLS data involved.

 I switched to a `&[(WString, Vec<WString>)]`, it's nice now because the variables will be in a deterministic order.